oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenIntrospectionEndpointFilter.java

```diff
@@ -0,0 +1,450 @@
+	}
+
+	@Override
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+			throws ServletException, IOException {
+
+		if (!this.tokenEndpointMatcher.matches(request)) {
+			filterChain.doFilter(request, response);
+			return;
+		}
+
+		// obtain Authentication
+		Authentication principal = SecurityContextHolder.getContext().getAuthentication();
+		if (!isPrincipalAuthenticated(principal)) {
+			// Pass through the chain with the expectation that the authentication process
+			// will commence via AuthenticationEntryPoint
+			filterChain.doFilter(request, response);
+			return;
+		}
+
+		MultiValueMap<String, String> parameters = OAuth2EndpointUtils.getParameters(request);
+		try {
+			// token (REQUIRED)
+			String token = parameters.getFirst(OAuth2ParameterNames2.TOKEN);
+			if (!StringUtils.hasText(token) || parameters.get(OAuth2ParameterNames2.TOKEN).size() != 1) {
+				throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames2.TOKEN);
+			}
+
+			// token_type_hint (OPTIONAL)
+			Optional<TokenType> tokenTypeHint = Optional
+					.ofNullable(parameters.getFirst(OAuth2ParameterNames2.TOKEN_TYPE_HINT)).map(TokenType::new);
+			if (tokenTypeHint.isPresent() && parameters.get(OAuth2ParameterNames2.TOKEN_TYPE_HINT).size() != 1) {
+				throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames2.TOKEN_TYPE_HINT);
+			}
+			TokenIntrospectionSuccessResponse tokenIntrospectionResponse = generateTokenIntrospectionResponse(principal,
+					token, tokenTypeHint);
+
+			this.sendTokenIntrospectionResponse(response, tokenIntrospectionResponse);
+		} catch (OAuth2AuthenticationException ex) {
+			SecurityContextHolder.clearContext();
+			sendErrorResponse(response, ex.getError());
+		}
+	}
+
+	private TokenIntrospectionSuccessResponse generateTokenIntrospectionResponse(Authentication principal, String token,
+			Optional<TokenType> tokenTypeHint) {
+		try {
+			// get client from token
+			OAuth2Authorization authorization = findAuthorizationByToken(token, tokenTypeHint)
+					.orElseThrow(() -> new InvalidTokenException("Token not found"));
+
+			// check if token corresponds to authorized Client
+			OAuth2ClientAuthenticationToken clientAuthentication = principal instanceof OAuth2ClientAuthenticationToken
+					? (OAuth2ClientAuthenticationToken) principal
+					: null;
+			if (clientAuthentication == null || !clientAuthentication.getRegisteredClient().getId()
+					.equals(authorization.getRegisteredClientId())) {
+				throw new InvalidTokenException("Token does not correspond to authenticated client");
+			}
+
+			// we obtained the authorization from the token, no need to validate Optional here
+			AbstractOAuth2Token oauthToken = authorization.getTokens().getToken(token).get();
+			if (authorization.getTokens().getTokenMetadata(oauthToken).isInvalidated()) {
+				throw new InvalidTokenException("Token has been invalidated");
+			}
+
+			oauthToken = parseJwt(oauthToken);
```

Comment:
I first thought of validating both the _AbstractOAuth2Token_ implementation (which could be an _OAuth2AccessToken_, an _OAuth2RefreshToken_, or an _OidcIdToken_) and the _Jwt_ separately, but according to the following specs:
https://tools.ietf.org/html/draft-ietf-oauth-access-token-jwt-10
https://openid.net/specs/openid-connect-core-1_0.html#IDToken
the jwt-formatted token counterparts **will** contain ( according to their required fields) all the base token fields (mainly _iat_, _exp_) so I decided to simply validate it as a _Jwt_ if possible, and avoid validating things twice. I couldn't find a spec for the JWT profile for a _refresh_token_.